src/AutoSchedule.cpp

```diff
@@ -0,0 +1,2312 @@
+        regions[out.name()] = out_box;
+
+        for (auto &reg: regions) {
+            // Merge region with an existing region for the function in the global map
+            if (pipeline_bounds.find(reg.first) == pipeline_bounds.end())
+                pipeline_bounds[reg.first] = reg.second;
+            else
+                merge_boxes(pipeline_bounds[reg.first], reg.second);
+        }
+    }
+
+    return pipeline_bounds;
+}
+
+struct Partitioner {
+
+    struct FusionChoice {
+        // FusionChoice encodes the choice of the prod_group being merged with
+        // the cons_group at the granularity of the tile given by tile_sizes
+        string prod;
+        FStage cons;
+
+        FusionChoice(string _prod, FStage _cons) : prod(_prod), cons(_cons) {}
+
+        bool operator==(const FusionChoice &other) const {
+            return (prod == other.prod) &&
+                    (cons == other.cons);
+        }
+
+        bool operator<(const FusionChoice &other) const {
+            return prod < other.prod || (prod == other.prod &&
+                                         cons < other.cons) ;
+        }
+
+        friend std::ostream& operator<<(std::ostream &stream,
+                                        const FusionChoice &choice) {
+            stream << "Choice:" << choice.prod << "->" << choice.cons << '\n';
+            return stream;
+        }
+    };
+
+    struct Group {
+        // The output stage representing the group
+        FStage output;
+        // All the functions that belong to the group
+        vector<FStage> members;
+
+        // Schedule information
+        // All the members of the group which are inlined
+        set<string> inlined;
+        // Tile sizes along the dimensions of the output function of the group
+        map<string, int> tile_sizes;
+
+        Group(FStage _output, vector<FStage> _members):
+              output(_output), members(_members) { }
+
+        friend std::ostream& operator <<(std::ostream &stream, const Group &g) {
+
+            stream << "Output FStage:" << g.output << '\n';
+            stream << "Members:" << '[';
+            for (auto &m: g.members) {
+                stream << m << ",";
+            }
+            stream << "]" << '\n';
+
+            stream << "Inlined:" << '[';
+            for (auto &in: g.inlined) {
+                stream << in << ",";
+            }
+            stream << "]" << '\n';
+
+            stream << "Tile sizes:" << "[";
+            for (auto &s: g.tile_sizes) {
+                stream << "(" << s.first << "," <<  s.second << ")";
+            }
+            stream << "]" << '\n';
+
+            return stream;
+        }
+    };
+
+    struct GroupAnalysis {
+        // Estimate of arithmetic cost
+        int64_t arith_cost;
+        // Estimate of accesses to slow memory
+        int64_t mem_cost;
+        // Estimate of the parallelism
+        int64_t parallelism;
+
+        friend std::ostream& operator <<(std::ostream &stream,
```

Comment:
I think we always prefer no space between 'operator' and '<<' (for all operator overloads)... see IROperator.h